File: BaekJoon/BackTracking/숫자판점프_2210.cpp
//
//  숫자판점프_2210.cpp
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/01/29.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: 숫자판점프_2210

// MARK: - DFS 문제풀이
#if 0
#include <iostream>
#include <vector>
#include <set>
using namespace std;

vector<vector<int>> G(5, vector<int>(5,0));
vector<int> V;
set<vector<int>> ST;
int tx[] = {0,0,-1,1};
int ty[] = {-1,1,0,0};

void DFS(int x, int y) {
    if(V.size() > 6) return;
    if(V.size() == 6) {
        ST.insert(V);
        return;
    }
    
    for(int i=0; i<4; i++) {
        int nx = x + tx[i];
        int ny = y + ty[i];
        if(nx < 0 || ny < 0 || nx >= 5 || ny >= 5) continue;
        V.push_back(G[nx][ny]);
        DFS(nx, ny);
        V.pop_back();
    }
}

int main() {
    ios_base :: sync_with_stdio(0); cin.tie(0);
    for(int i=0; i<5; i++) {
        for(int j=0; j<5; j++) {
            cin >> G[i][j];
        }
    }
    
    for(int i=0; i<5; i++) {
        for(int j=0; j<5; j++) {
            V = {G[i][j]};
            DFS(i,j);
        }
    }
    
    printf("%d\n",(int)ST.size());
    return 0;
}
#endif

File: BackJoonPS/BackjoonAlgorithmPS/BackTracking/숫자판점프_2210.swift
//
//  숫자판점프_2210.swift
//  BackJoonPractice
//
//  Created by MinKyeongTae on 2020/01/30.
//  Copyright © 2020 Min Kyeong Tae. All rights reserved.
//

// MARK: 숫자판점프_2210

// MARK: - DFS 문제풀이

/*
 import Foundation

 var G = [[String]](repeating: [String](), count: 5);
 var MP = [String:Int]()
 var tx = [0,0,-1,1]
 var ty = [-1,1,0,0]
 var Ans = 0, len = 5

 func DFS(_ cnt: Int, _ S: String, _ X: Int, _ Y: Int) {
 if cnt >= 6 { MP[S] = 1; return }
 for i in tx.indices {
     let nx = X + tx[i]
     let ny = Y + ty[i]
     if nx < 0 || ny < 0 || nx >= len || ny >= len { continue }
     DFS(cnt+1, S+G[nx][ny], nx, ny)
 }
 }

 for i in G.indices {
 let arr = readLine()!.split(separator: " ").map { String($0) }
 G[i] = arr;
 }

 for i in G.indices {
 for j in G[i].indices {
     DFS(1, G[i][j], i, j)
 }
 }

 print(MP.count)
 */
